### PR TITLE
Add some edge cases

### PR DESCRIPTION
**Pull request by hax**
Note the last two new cases are fail now.

There are also some cases which I am not sure which is the "right" solution, pls see the comments.

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -13,6 +13,15 @@ describe('Remove', function () {
     expect(semi.remove(src)).to.equal('var a = 123\na++\n')
   })
 
+  // Though semi can be omitted in this case, it's definitly not a good coding style.
+  // Two options:
+  // 1. throw a warning
+  // 2. do some reformat, such as replace semi with a newline?
+  it('newline within multiline comment', function () {
```
Comment (hax) [inline test/test.js:20]: or 3. just ignore this edge case :)

File: test/test.js [hunk 2]
```diff
@@ -46,6 +55,76 @@ describe('Remove', function () {
...
+    var src = "for (;;);\n"
+    expect(semi.remove(src)).to.equal(src)
+
+    // for...in
+    var src = "for (var key in obj);\n"
+    expect(semi.remove(src)).to.equal(src)
+  })
+
+  // The better coding style for these cases should be add {} ,
+  // but I'm not sure whether it is the duty of this project
```
Comment (hax) [inline test/test.js:85]: If we have another component which will wrap only statement with {} , we could put it before semi in the process chain.

File: test/test.js [hunk 3]
```diff
@@ -46,6 +55,76 @@ describe('Remove', function () {
...
+    expect(semi.remove(src)).to.equal(src)
+
+  })
+
+  it('do...while', function () {
+    // should remove semi
+    var src = "do { x-- } while (x);\n"
+    expect(semi.remove(src)).to.equal(src.replace(/;/g, ''))
+
+    // should add semi
```
Comment (hax) [inline test/test.js:118]: fail now

File: test/test.js [hunk 4]
```diff
@@ -46,6 +55,76 @@ describe('Remove', function () {
...
+    var src = "do { x-- } while (x)\n+x"
+    expect(semi.remove(src)).to.equal(src.replace(/\n/, '\n;'))
+  })
+
+  it('var statement', function () {
+    // should add semi
```
Comment (hax) [inline test/test.js:124]: fail now.

This is a very interesting case which I learned from the source code of seajs https://github.com/seajs/seajs/blob/master/src/standalone.js#L2-L3
cc @army8735

File: test/test.js [hunk 5]
```diff
@@ -88,4 +167,4 @@ describe('Add', function () {
     expect(semi.add(src)).to.equal('a++;\n/a/.test(b);')
   })
 
-})
```
Comment (hax) [inline test/test.js:91]: Oops, NewLine before EOF :)

File: test/test.js (hunk 1 above)
Comment (yyx990803) [inline test/test.js:20]: just ignore it, since this project doesn't aim to help fix styles issues other than semicolons.

File: test/test.js [hunk 6]
```diff
@@ -46,6 +55,76 @@ describe('Remove', function () {
...
+
+  })
+
+  it('do...while', function () {
+    // should remove semi
+    var src = "do { x-- } while (x);\n"
+    expect(semi.remove(src)).to.equal(src.replace(/;/g, ''))
+
+    // should add semi
+    var src = "do { x-- } while (x)\n+x"
```
Comment (yyx990803) [inline test/test.js:119]: Will not pass this one because it is also valid JavaScript, and we don't want to assume we know what the user actually wants.

Comment (yyx990803) [inline test/test.js:119]: I'm debating whether we should add semicolons to already valid JavaScript - on one hand it's about making the style consistent; on the other hand it's already valid and working.

**Comment by yyx990803** [comment]
Thanks a lot! I did a refactor using ESLint instead of JSHint, which is much more customizable and offers actual AST info. Most of these tests are now included except for two special cases:
1. adding semicolon to the next line after `do...while`.
2. adding semicolon to the next line after a `var` declaration.

It's interesting because for these two statements, even if the next line starts with `[\/\(\[\+\-]`, they are still valid JavaScript and work as intended. I'm not sure if re-formatting already valid JS is within the responsibility of this project, especially when executing "remove semicolons" actually adds semicolons... which can be quite confusing. So I left these out intentionally (they are actually implemented, just commented out.) I'd be happy to hear your thoughts on it.

File: test/test.js (hunk 4 above)
Comment (army8735) [inline test/test.js:124]: haha

**Comment by hax** [comment]
@yyx990803 
Actually there are not only two such cases. All var/function/let/generator/class declaration suffer the same problem. I've written more test cases these days, resulting 39 passing/24 failing as your JSHint based implementation. I will try your latest implementation soon.

And yes, it seems weird to 'remove' actually add semis, but I think it's a API naming problem (change semi.remove/add to semi.less/more?). The key point of the project is to transform code from classic style to semicolon-less style. So adding extra semicolon is ok because it's more like the code written by hand. And the consistence make future editing more easy (eg. copy and paste such lines).

**Comment by hax** [comment]
Just tested new impl, now 43 passing/20 failing :)  All fail cases are adding leading ; for already valid code except fail case 6.

```
1) should prepend leading semicolon to line starts with unary - operator:
"-x" => ";-x"
2) should prepend leading semicolon to line starts with unary + operator:
"+x" => ";+x"
3) should prepend leading semicolon to line starts with regexp literal:
"/x/" => ";/x/"
4) should prepend leading semicolon to line starts with array initializer:
"[x]" => ";[x]"
5) should prepend leading semicolon to line starts with grouping operator:
"(x)" => ";(x)"
6) should remove redundant semicolon after for-of:
"for (value of obj) {};" => "for (value of obj) {}"
7) should prepend leading semicolon after block:
"{}\n-x" => "{}\n;-x"
8) should prepend leading semicolon after if:
"if (test) {}\n-x" => "if (test) {}\n;-x"
9) should prepend leading semicolon after else:
"if (test) {} else {}\n-x" => "if (test) {} else {}\n;-x"
10) should prepend leading semicolon after while:
"while (test) {}\n-x" => "while (test) {}\n;-x"
11) should prepend leading semicolon after for:
"for (;;) {}\n-x" => "for (;;) {}\n;-x"
12) should prepend leading semicolon after for-in:
"for (key in obj) {}\n-x" => "for (key in obj) {}\n;-x"
13) should prepend leading semicolon after for-of:
"for (value of obj) {}\n-x" => "for (value of obj) {}\n;-x"
14) should prepend leading semicolon after do...while:
"do {} while (i--)\n-x" => "do {} while (i--)\n;-x"
15) should prepend leading semicolon after var declaration:
"var a, b\n-x" => "var a, b\n;-x"
16) should prepend leading semicolon after function declaration:
"function f() {}\n-x" => "function f() {}\n;-x"
17) should prepend leading semicolon after let declaration:
"let a, b\n-x" => "let a, b\n;-x"
18) should prepend leading semicolon after generator declaration:
"function *g() {}\n-x" => "function *g() {}\n;-x"
19) should prepend leading semicolon after class declaration:
"class c {}\n-x" => "class c {}\n;-x"
20) should warn and prepend semicolon after comment span multiple lines:
"x /* ...\n */ -x" => "x /* ...\n */ ;-x"
```

Here is my testings: https://github.com/hax/semicolon-less.js

**Comment by yyx990803** [comment]
Implemented a branch which always adds leading semicolons: https://github.com/yyx990803/semi/tree/format

Let me know how it goes with the additional tests.

**Comment by hax** [comment]
Cool! 58 passing and only 5 failing.

```
1) should remove redundant semicolon after for-of:
"for (value of obj) {};" => "for (value of obj) {}"
2) should prepend leading semicolon after for-of:
"for (value of obj) {}\n-x" => "for (value of obj) {}\n;-x"
3) should prepend leading semicolon after generator declaration:
"function *g() {}\n-x" => "function *g() {}\n;-x"
4) should prepend leading semicolon after class declaration:
"class c {}\n-x" => "class c {}\n;-x"
5) should warn and prepend semicolon after comment span multiple lines:
"x /* ...\n */ -x" => "x /* ...\n */ ;-x"
```

The first 4 cases are ES6 features.
The last case is only theoretical.
So I think it's near perfection!

**Comment by yyx990803** [comment]
Nice, for ES6 we will have to wait for ESLint to support it (it has an experimental branch using esprima-fb, but may eventually depend on official ES6 support in Esprima.)

**Comment by yyx990803** [comment]
Added `leading` option for this behavior:

``` js
semi.remove(src, { leading: true }) // always add leading semicolons
```

From command line:

``` bash
semi rm src.js --leading
```

**Comment by hax** [comment]
Could `leading` option default to true?

Even keep it default to false, there is a common case -- IIFE which need special treatment:

``` js
(function () {
...
}());
```

If we just remove the trailing semicolon, concat of multiple js may be break.

**Comment by hax** [comment]
close this and will open new issue for remain problems
